### PR TITLE
TOB-ALEPH-007 (Errors occurring in async code leave program in inconsistent state).

### PR DESCRIPTION
**Pull request by mike1729**

### CODE REVIEW

File: src/testing/mock.rs [hunk 1]
```diff
@@ -182,26 +182,24 @@ impl Future for HonestHub {
...
     }
-}
 
-impl Spawner {
-    pub(crate) async fn wait(&self) {
-        for h in self.handles.lock().iter_mut() {
-            let _ = h.await;
-        }
-    }
-    pub(crate) fn new() -> Self {
```
Comment (timorl) [inline src/testing/mock.rs:201]: I would suggest leaving the `new` even if it is trivial, future changes are likely to be simpler (just look at the files in which you had to change only this).

File: src/lib.rs [hunk 2]
```diff
@@ -93,6 +93,13 @@ pub type OrderedBatch<Data> = Vec<Data>;
 pub trait SpawnHandle: Clone + Send + 'static {
     /// Run a new task.
     fn spawn(&self, name: &'static str, task: impl Future<Output = ()> + Send + 'static);
+    /// Run a new task and returns a handle to it. If there is some error or panic during
+    /// execution of the task, the handle should return an error.
+    fn spawn_essential(
+        &self,
+        name: &'static str,
+        task: impl Future<Output = ()> + Send + 'static,
+    ) -> Pin<Box<dyn Future<Output = Result<(), ()>> + Send>>;
```
Comment (timorl) [inline src/lib.rs:102]: Maybe some type alias for this?

**Review by timorl — CHANGES_REQUESTED** [review]
Minor comments, although in light of our earlier discussion I'm not sure what the ultimate fate of this PR should be.

File: src/testing/mock.rs (hunk 1 above)
Comment (mike1729) [reply src/testing/mock.rs:201]: Done.

**Review by timorl — APPROVED** [review]

**Review by timorl — APPROVED** [review]

**Pull request merged**